src/pow.rs

```diff
@@ -0,0 +1,1540 @@
+        } else {
+            d.low_u128()
+        }
+    }
+}
+do_impl!(Target);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Big-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+struct Uint256(pub [u128; 2]);
+
+impl Uint256 {
+    const MAX: Uint256 = Uint256([
+        0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff,
+        0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff,
+    ]);
+
+    const ZERO: Uint256 = Uint256([0, 0]);
+
+    const ONE: Uint256 = Uint256([0, 1]);
+
+    /// Creates [`Uint256`] from a big-endian array of `u8`s.
+    fn from_be_bytes(a: [u8; 32]) -> Uint256 {
+        let mut out = [0u128; 2];
+        out.iter_mut().zip(a.chunks_exact(16)).for_each(|(word, bytes)| {
+            *word = {
+                let mut owned = [0_u8; 16];
+                owned.copy_from_slice(bytes);
+                u128::from_be_bytes(owned)
+            }
+        });
+        Uint256(out)
+    }
+
+    /// Creates a [`Uint256`] from a little-endian array of `u8`s.
+    fn from_le_bytes(a: [u8; 32]) -> Uint256 {
+        let mut out = [0u128; 2];
+        out.iter_mut().zip(a.chunks_exact(16).rev()).for_each(|(word, bytes)| {
+            *word = {
+                let mut owned = [0_u8; 16];
+                owned.copy_from_slice(bytes);
+                u128::from_le_bytes(owned)
+            }
+        });
+        Uint256(out)
+    }
+
+    /// Converts `Self` to a big-endian array of `u8`s.
+    fn to_be_bytes(self) -> [u8; 32] {
+        let Uint256(ref arr) = self;
+        let mut res = [0; 32];
+        for (i, word) in arr.iter().enumerate() {
+            let start = i * 16;
+            res[start..start + 16].copy_from_slice(&word.to_be_bytes());
+        }
+        res
+    }
+
+    /// Converts `Self` to a little-endian array of `u8`s.
+    fn to_le_bytes(self) -> [u8; 32] {
+        let Uint256(ref arr) = self;
+        let mut res = [0; 32];
+        for (i, word) in arr.iter().rev().enumerate() {
+            let start = i * 16;
+            res[start..start + 16].copy_from_slice(&word.to_le_bytes());
+        }
+        res
+    }
+
+    /// Calculates 2^256 / (x + 1) where x is a 256 bit unsigned integer.
+    ///
+    /// 2**256 / (x + 1) == ~x / (x + 1) + 1
+    ///
+    /// (Equation shamelessly stolen from bitcoind)
+    fn inverse(&self) -> Uint256 {
+        // We should never have a target/work of zero so this doesn't matter
+        // that much but we define the inverse of 0 as max.
+        if self.is_zero() {
+            return Uint256::MAX;
+        }
+        // We define the inverse of 1 as max.
+        if self.is_one() {
+            return Uint256::MAX;
+        }
+        // We define the inverse of max as 1.
+        if self.is_max() {
+            return Uint256::ONE;
+        }
+
+        let ret = !*self / self.wrapping_inc();
+        ret.wrapping_inc()
+    }
```

Comment:
When making this const fn we can't trivially use the array trick because of possible division by zero. A workaround:

```rust
let mapped = [*self, Self::ZERO][self.is_max() as usize];
```

then work with `mapped` instead of self and conditionally return the appropriate value at the end.